freeciv-web/src/main/java/org/freeciv/services/Validation.java

```diff
@@ -0,0 +1,42 @@
+/*******************************************************************************
+ * Freeciv-web
+ * Copyright (C) 2018 The Freeciv-web project
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *******************************************************************************/
+package org.freeciv.services;
+
+import java.util.regex.Pattern;
+
+public class Validation {
+
+	private static Pattern usernamePattern = Pattern.compile("[a-z][a-z0-9]*");
+
+	public static boolean isValidUsername(String name) {
```

Comment:
> pure function, with the class just a class providing a kind of namespace, without any state

All true, but not relevant to testability.

`isValidUsername(...)` is simple, but has 5 branches. Yesterday, passing "myname1" would produce `false`, but today `true`. In the future, other changes may be made.

With that in mind, when testing a method (e.g. `MyClass.testMe(...)`) that _uses_ `Validation.isValidUsername(...)`, you want to test cases where `isValidUsername(...)` passes, and where it fails within `testMe(...)` (and maybe when it throws an exception) - so if you _must_ use the real `isValidUsername(...)`, you also must know what input will cause it produce those results. In short, you're testing `testMe(...)`, and implicitly `isValidUsername(...)` also, making the tests more [complicated](http://noop.nl/2008/08/simple-vs-complicated-vs-complex-vs-chaotic.html).

To avoid that, we mock `isValidUsername(...)` - use a fake version of the method that simply does what we want during tests, regardless of input. Changing implementation of a static method [is possible](https://github.com/powermock/powermock/wiki/Motivation) (with behavioral edge-cases due to their "global" nature), but not recommended - change it once you're changing it everywhere. You need to change it _back_ to the original, etc... With instances, the instance mocked can be different in each test, avoiding the problems of static methods, and insuring the changes don't affect other tests (or worse, the test framework).

Assuming we make `isValidUsername(...) an instance method, how does `MyClass` gets an instance of `Validation` - as a mock (during tests) or otherwise (at runtime)? The "[Dependency Injection](https://en.wikipedia.org/wiki/Dependency_injection)" pattern (aka. "inversion of control", "IOC", etc). Using DI, during our tests we create a mock `Validation` instance, and for each test of `testMe(...)` tell the `validation` mock how `isValidUsername(...)` should behave. Here is a simple Java test using [Mockito](https://site.mockito.org/).

Consider this JUnit / [Mockito](https://site.mockito.org/) based Java test, where `testMethod(...)` changes state if the argument is valid (worth noting that annotations be used to reduce the code more, but I thought this would be more clear to those unfamiliar w/ mockito):

    public class MyClassTest {
        Validation mockValidation;
        MyClass sut;

        @Before
        void setup() {
            mockValidation = mock(Validation.class);
            sut = new MyClass(mockValidation); // constructor injection (preferred) though setter methods work
        }
        @Test
        void testMe_withValidUsername_changesState() {
            when(mockValidation.isValidUserName(any())).thenReturn(true); // regardless of input, return true

            myClass.testMe("don't care at all");

            assertThat(myClass.getState(), equalTo(expectedStateChanged)); // verify behavior of `testMe`
        }
        // testMe_withInvalidValidUsername_noStateChange...
    }

As you point out, `isValidUsername(name)` is a pure function without state. Image a method  `calculateMyTaxReturnPayment(x,y,z,p,d,q,r,o,f,l,m,a,O)` that is also a pure function and stateless... as is `String.copyValueOf(charArray)`. _Really_ simple utility functions (e.g. language runtime API like the `String` statics), you clearly _can't_ change, and wouldn't want to anyway. Some application utilities are so simple, they don't bear being injected (e.g. `MyStringUtils.firstLetter(...)`). So how to decide what needs to be testable? It's a judgement call, but a good rule of thumb is that if there is _any_ business logic, it should be testable, (username validity is clearly business logic, given it is evolving). I generally err on the side of caution, with the "trust me you'll regret it..." warning.

Finally, in order to use DI, a framework to create all objects in the application and "wire them up" is needed. [Guice](https://github.com/google/guice) is a good, lightweight one, you may have heard of [Spring](https://spring.io/) which has DI (and much more), and there are others. We don't have DI (yet!) in freeciv-web, though, again, erring on the side of caution, I prefer to follow the pattern as much as possible to reduce refactoring later. Thus I suggest:

1. Design the classes as testable, instantiating the service (Validation) in the constructor of classes where used.
2. Incorporate a DI framework, making only small modification to the constructors.
3. Add tests

2. and 3. could be swapped - or never happen, 1. is still a good choice regardless.

Reply:
> apologies for the length

On the contrary, thank you for taking the time and effort to write it.

While I'm not still a convert, there's enough to convince me to change, with quite a few things I had not even considered before, and surely a lot more when all that sinks in and more practice backs it up.